miniml/compiler/compile.scm

```diff
@@ -562,49 +607,33 @@
-        ((char=? c #\+) (make-lexical-token 'PLUS location #f))
-        ((char=? c #\-) (if (char=? (peek-char) #\>)
-                            (begin (read-char) (make-lexical-token 'MINUSGT location #f))
-                            (if (char-numeric? (peek-char))
-                                (make-lexical-token 'INT location (- (string->number (list->string (number-chars errorp)))))
-                                (make-lexical-token 'MINUS location #f))))
-        ((char=? c #\*) (make-lexical-token 'STAR location #f))
-        ((char=? c #\~) (make-lexical-token 'TILDE location #f))
-        ((char=? c #\@) (cond
-                         ((char=? (peek-char) #\@)
-                          (read-char) (make-lexical-token 'ATAT location #f))
-                         (else (make-lexical-token 'AT location #f))))
-        ((char=? c #\^) (make-lexical-token 'CARET location #f))
-        ((char=? c #\?) (make-lexical-token 'QUESTION location #f))
-        ((char=? c #\!) (make-lexical-token 'BANG location #f))
-        ((char=? c #\&) (if (char=? (peek-char) #\&)
-                            (begin (read-char) (make-lexical-token 'AMPERAMPER location #f))
-                            (errorp "Illegal character: " c)))
-        ((char=? c #\<) (if (char=? (peek-char) #\>)
-                            (begin (read-char) (make-lexical-token 'LTGT location #f))
-                            (if (char=? (peek-char) #\=)
-                                (begin (read-char) (make-lexical-token 'LTEQ location #f))
-                                (if (char=? (peek-char) #\-)
-                                    (begin (read-char) (make-lexical-token 'LTMINUS location #f))
-                                    (make-lexical-token 'LT location #f)))))
-        ((char=? c #\>) (if (char=? (peek-char) #\=)
-                            (begin (read-char) (make-lexical-token 'GTEQ location #f))
-                            (make-lexical-token 'GT location #f)))
+        ((char=? c #\+) (mksymbol location errorp c))
+        ((char=? c #\-) (if (char-numeric? (peek-char))
+                            (make-lexical-token 'INT location (- (string->number (list->string (number-chars errorp)))))
+                            (mksymbol location errorp c)))
+        ((char=? c #\*) (mksymbol location errorp c))
+        ((char=? c #\~) (mksymbol location errorp c))
+        ((char=? c #\@) (mksymbol location errorp c))
+        ((char=? c #\^) (mksymbol location errorp c))
+        ((char=? c #\?) (mksymbol location errorp c))
+        ((char=? c #\!) (mksymbol location errorp c))
+        ((char=? c #\&) (mksymbol location errorp c))
+        ((char=? c #\<) (mksymbol location errorp c))
+        ((char=? c #\>) (mksymbol location errorp c))
+        ((char=? c #\/) (mksymbol location errorp c))
+        ((char=? c #\%) (mksymbol location errorp c))
+        ((char=? c #\$) (mksymbol location errorp c))
```

Comment:
It looks a bit painful to list all these starting-symbols with a copy of the semantic action for each one. Naively one would think of grouping them with a `starting-symbol?` predicate. I suppose that I may be missing something related to condition ordering that makes it not work / not work in a regular way.

Reply:
I grouped them as well. However, I didn't use a `starting-symbol?` predicate as it would be misleading, since `'|'` and `'-'` have to be special-cased.